Held item refactor

Wlowscha opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-       [Bug]: If the PR is primarily a bug fix
-      [Move]: If a move has new or changed functionality
-   [Ability]: If an ability has new or changed functionality
-      [Item]: For new or modified items
-   [Mystery]: For new or modified Mystery Encounters
-      [Test]: If the PR is primarily adding or modifying tests
-     [UI/UX]: If the PR is changing UI/UX elements
-     [Audio]: If the PR is adding or changing music/sfx
-    [Sprite]: If the PR is adding or changing sprites
-   [Balance]: If the PR is related to game balance
- [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-      [Docs]: If the PR is just adding or modifying documentation (such as tsdocs/code comments)
-    [GitHub]: For changes to GitHub workflows/templates/etc
-      [Misc]: If no other category fits the PR
-->
<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small)
-->

## What are the changes the user will see?
<!-- Summarize what are the changes from a user perspective on the application -->

## Why am I making these changes?
<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->

## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->

## How to test the changes?
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->

## Checklist
- [ ] **I'm using `beta` as my base branch**
- [ ] There is no overlap with another PR?
- [ ] The PR is self-contained and cannot be split into smaller PRs?
- [ ] Have I provided a clear explanation of the changes?
- [ ] Have I tested the changes manually?
- [ ] Are all unit tests still passing? (`npm run test`)
  - [ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [ ] Have I provided screenshots/videos of the changes (if applicable)?
  - [ ] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [ ] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [ ] If so, please leave a link to it here: 
- [ ] Has the translation team been contacted for proofreading/translation?

Wlowscha commented:
https://github.com/user-attachments/assets/c08b367c-b79a-4db3-bdaf-fefd6dc8e4c4

Documenting progress: HeldItems now fully functional, although still a work in progress.

Instead of rewriting everything, the new HeldItem logic is being coupled with the old ModifierType logic for rewards (which will slowly be phased out).

The video shows the two separate instances of Leftover rewards. The new instance does not yet show up in the summary, but works in game. The old instance does not work anymore. Tests have not yet been updated, so the Leftovers test appropriately fails.

SirzBenjie commented on src/enums/held-items.ts:
```diff
@@ -0,0 +1,88 @@
+type HeldItemValue = typeof HeldItems[HeldItemName];
+
+// Use a type-safe reducer to force number keys and values
+export const HeldItemNames: Record<HeldItemValue, HeldItemName> = Object.entries(HeldItems).reduce(
+  (acc, [key, value]) => {
+    acc[value as HeldItemValue] = key as HeldItemName;
+    return acc;
+  },
+  {} as Record<HeldItemValue, HeldItemName>
+);
```
I really like this.

SirzBenjie commented on src/modifier/modifier-type.ts:
```diff
@@ -1328,6 +1399,63 @@ export class FusePokemonModifierType extends PokemonModifierType {
+        return new AttackTypeBoosterReward(pregenArgs[0] as PokemonType, TYPE_BOOST_ITEM_BOOST_PERCENT);
+      }
+
+      const attackMoveTypes = party.flatMap(p =>
+        p
+          .getMoveset()
+          .map(m => m.getMove())
+          .filter(m => m instanceof AttackMove)
+          .map(m => m.type),
```
Can you put a TODO comment here that makes it consider the type of variable-type moves / abilities
This would let something like a mon with refrigerate and normal moves allow nevermeltice to generate.

Wlowscha replied:
I guess what we would need is a method on Pokemon which takes the move as input and returns its type. Unless more information is needed which is not on the Pokemon... How does Plasma Fists work, for example? Although it would not be relevant here specifically.

SirzBenjie replied:
I can worry about the implementation, I just want a comment here, as it's a place that wants to know about the move type.

emdeann replied:
Late to the party here lol but `Pokemon#getMoveType` does already exist and considers everything including plasma fists

SirzBenjie commented on src/modifier/all-held-items.ts:
```diff
@@ -0,0 +1,224 @@
+    return this;
+  }
+
+  getMaxStackCount(): number {
+    return this.maxStackCount;
+  }
```
Are we able to make this into a getter, or, better yet, have maxStackCount just be a readonly property that doesn't need this method?

I think this is mirroring the existing logic; can we fix it?

SirzBenjie commented on src/modifier/all-held-items.ts:
```diff
@@ -0,0 +1,224 @@
+  createIcon(stackCount: number, _forSummary?: boolean): Phaser.GameObjects.Container {
+    const container = globalScene.add.container(0, 0);
+
+    const item = globalScene.add.sprite(0, 12, "items");
+    item.setFrame(this.getIcon());
+    item.setOrigin(0, 0.5);
```
If you merge beta into this branch, which has the fixes to mocks from #5696 you can (and should) use phaser chaining

```ts
const item = globalScene.add.sprite(0, 12, "items").setFrame(this.getIcon()).setOrigin(0, 0.5);
```

SirzBenjie commented on src/modifier/all-held-items.ts:
```diff
@@ -0,0 +1,224 @@
+    const text = globalScene.add.bitmapText(10, 15, "item-count", stackCount.toString(), 11);
+    text.letterSpacing = -0.5;
+    if (stackCount >= this.getMaxStackCount()) {
+      text.setTint(0xf89890);
```
I don't like magic numbers. Can we make this a constant somewhere? (Is it not already)?

SirzBenjie commented on src/modifier/all-held-items.ts:
```diff
@@ -0,0 +1,224 @@
+export function initHeldItems() {
+  // SILK_SCARF, BLACK_BELT, etc...
+  for (const [typeKey, heldItemType] of Object.entries(attackTypeToHeldItem)) {
+    const pokemonType = Number(typeKey) as PokemonType;
```
The call to `Number` here seems fishy. I'm sure there's a way to make typescript not find this necessary.

SirzBenjie commented on src/modifier/modifier-type.ts:
```diff
@@ -1328,6 +1399,63 @@ export class FusePokemonModifierType extends PokemonModifierType {
+
+      const attackMoveTypeWeights = new Map<PokemonType, number>();
+      let totalWeight = 0;
+      for (const t of attackMoveTypes) {
+        if (attackMoveTypeWeights.has(t)) {
+          if (attackMoveTypeWeights.get(t)! < 3) {
+            // attackMoveTypeWeights.has(t) was checked before
+            attackMoveTypeWeights.set(t, attackMoveTypeWeights.get(t)! + 1);
+          } else {
+            continue;
+          }
+        } else {
+          attackMoveTypeWeights.set(t, 1);
+        }
+        totalWeight++;
```
```suggestion
      for (const t of attackMoveTypes) {
        const weight = attackMoveTypeWeights.get(t) ?? 0;
        if (weight < 3) {
          attackMoveTypeWeights.set(t, weight + 1)
          totalWeight++
        }
```

Shorter way of writing the same thing.

DayKev commented on src/modifier/all-held-items.ts:
This file seems to have a bunch of different things in it. Imo it should be broken apart somewhat.

Wlowscha commented:
![newHeldItemSummary](https://github.com/user-attachments/assets/0cebc4b5-a3ab-4a56-87c7-2f13fdc7cbc2)

Held items now showing up properly in summary :)

DayKev commented on src/modifier/all-held-items.ts:
```diff
@@ -0,0 +1,224 @@
+    this.icon = this.getIcon();
+  }
+
+  getName(): string {
+    return "";
+  }
```
There shouldn't be both a `name` field and a `getName` method if `getName` does not return `name`.

Also in that case you can just use a getter (`get name(): string { ... }`) so that `item.name` can be used instead of `item.getName()`.

SirzBenjie replied:
Yeah we should use getters

DayKev commented on src/modifier/modifier-bar.ts:
This is a UI file? It should be in `src/ui/` then.

Wlowscha replied:
That makes sense... I'll move it when I actually get to the point of refactoring it :)

SirzBenjie commented on src/items/held-items/attack-type-booster.ts:
```diff
@@ -3,8 +3,16 @@ import { PokemonType } from "#enums/pokemon-type";
+  /** The resolved type of the move */
+  moveType: PokemonType;
+  /** Holder for the damage value */
+  movePower: NumberHolder;
```
Doesn't need to be a number holder, a caller can just keep around the reference to the object it uses as params and then read it later on.

Wlowscha commented:
Changed several things in battle-scene.ts:
- Made a separate addHeldItem function which adds an item and updates modifiers in globalScene (still needs sounds)
- Also made a separate addFormChangeItem, which also incorporates FormChangeHeldItemModifier's obsolete .apply() method
- Removed held-item related logic from addModifier and similar functions
- Removed removePartyMemberModifiers (no need to erase modifiers since they are attached to the pokemon)
- Same applies to clearEnemyHeldItemModifiers
- Refactored tryTransferHeldItem and canTransferHeldItem (thanks to #5932 for much better understanding of the code)

Bertie690 commented on src/items/held-items/incrementing-stat.ts:
```diff
@@ -0,0 +1,61 @@
+
+export interface INCREMENTING_STAT_PARAMS {
+  /** The pokemon with the item */
+  pokemon: Pokemon;
+  stat: Stat;
+  statHolder: NumberHolder;
```
No need for a number holder if we're embedding it inside an object

Bertie690 commented on src/items/held-items/hit-heal.ts:
```diff
@@ -0,0 +1,53 @@
+  apply(params: HIT_HEAL_PARAMS): boolean {
+    const pokemon = params.pokemon;
+    const stackCount = pokemon.heldItemManager.getStack(this.type);
+    if (pokemon.turnData.totalDamageDealt && !pokemon.isFullHp()) {
```
```suggestion
    if (pokemon.turnData.totalDamageDealt > 0 && !pokemon.isFullHp()) {
```

Bertie690 reviewed (changes requested):
I know this is still a WIP, but just wanted to point out a few things that I saw.

DayKev commented on src/enums/held-item-id.ts:
I think the lower 2 bytes were supposed to be the category so that there would be room to increase the max number of items per category (in the unlikely event we need to do so).
```
SITRUS_BERRY: 0x0101,
LUM_BERRY: 0x0201,
ENIGMA_BERRY: 0x0301,
...

REVIVER_SEED: 0x0102,
WHITE_HERB: 0x0202,

SILK_SCARF: 0x0103,
BLACK_BELT: 0x0203,
...
```
